### PR TITLE
RSA support CRT and NO_CRT mode at same time

### PR DESCRIPTION
**Pull request by lubing521**
Keymaster VTS need libmbedtls with CRT in optee-os.
But some case of the Xtest 4006 no provide CRT
This will cause OP-tee OS check it fail. in order
to support this scenario work well,libmedtls need
frendly deal with RSA with CRT or not.

Signed-off-by: Lu Bing lubing@eswin.com
Tested-by: Lu Bing lubing@eswin.com

Notes:
* Pull requests cannot be accepted until the PR follows the [contributing guidelines](../CONTRIBUTING.md). In particular, each commit must have at least one `Signed-off-by:` line from the committer to certify that the contribution is made under the terms of the [Developer Certificate of Origin](../dco.txt).
* This is just a template, so feel free to use/remove the unnecessary things
## Description
A few sentences describing the overall goals of the pull request's commits.

## Status
**READY/IN DEVELOPMENT/HOLD**

## Requires Backporting
When there is a bug fix, it should be backported to all maintained and supported branches.
Changes do not have to be backported if:
- This PR is a new feature\enhancement
- This PR contains changes in the API. If this is true, and there is a need for the fix to be backported, the fix should be handled differently in the legacy branch

Yes | NO  
Which branch?

## Migrations
If there is any API change, what's the incentive and logic for it.

YES | NO

## Additional comments
Any additional information that could be of interest

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

## Steps to test or reproduce
Outline the steps to test or reproduce the PR here.

### CODE REVIEW

**Comment by yanesca** [comment]
@lubing521 Thank you very much for your contribution! Could you please help me understand why these changes are necessary?

Whether CRT is used in Mbed TLS  (the `MBEDTLS_RSA_NO_CRT` config option) is only an implementation detail, the computations should yield correct results either way. How do these test suites prescribe how the computation happens?

**Comment by lubing521** [comment]
> @lubing521 Thank you very much for your contribution! Could you please help me understand why these changes are necessary?
> 
> Whether CRT is used in Mbed TLS (the `MBEDTLS_RSA_NO_CRT` config option) is only an implementation detail, the computations should yield correct results either way. How do these test suites prescribe how the computation happens?

Thank you for your prompt reply.

The origin question as below
https://github.com/OP-TEE/optee_os/issues/4678

My current solution 
https://github.com/OP-TEE/optee_os/pull/4698

so  i try fix it on upstream.

**Comment by yanesca** [comment]
@lubing521 Thank you for the background information. I have read the thread, but it still does not explain why Keymaster VTS fails with `MBEDTLS_RSA_NO_CRT` on. That option supposed to be only a performance optimisation.

Could you please post the exact test cases that fail? (Input and expected output vs actual output.)

**Comment by lubing521** [comment]
> @lubing521 Thank you for the background information. I have read the thread, but it still does not explain why Keymaster VTS fails with `MBEDTLS_RSA_NO_CRT` on. That option supposed to be only a performance optimisation.
> 
> Could you please post the exact test cases that fail? (Input and expected output vs actual output.)

sorry，reply so later.

Keymaster TA in  function  TA_generate_key which on flie https://github.com/linaro-swg/kmgk/blob/master/keymaster/ta/generator.c
This function will create rsa key pair with CRT,so in optee-os with libmbedtls need enable CRT together.
if not ,this function will be called fail in rsa_check_context.
call stack as below in optee-os
syscall_asymm_operate--》crypto_acipher_rsanopad_decrypt–》mbedtls_rsa_private–》rsa_check_context

```
#if !defined(MBEDTLS_RSA_NO_CRT)
    /* Modular exponentiation for P and Q is only
     * used for private key operations and if CRT
     * is used. */
    if( is_priv &&
        ( mbedtls_mpi_cmp_int( &ctx->P, 0 ) <= 0 ||
          mbedtls_mpi_get_bit( &ctx->P, 0 ) == 0 ||
          mbedtls_mpi_cmp_int( &ctx->Q, 0 ) <= 0 ||
          mbedtls_mpi_get_bit( &ctx->Q, 0 ) == 0  ) )
    {
        return( MBEDTLS_ERR_RSA_BAD_INPUT_DATA );
    }
#endif /* !MBEDTLS_RSA_NO_CRT */
```
in other way, 
xtest_tee_test_4006 in  file  https://github.com/OP-TEE/optee_test/blob/master/host/xtest/regression_4000.c
these rsa  case will test diffrence way  with CRT or without CRT .

**Comment by yanesca** [comment]
Oh, I see, thank you for explaining.

**Comment by yanesca** [comment]
@lubing521 while we are waiting for reviews, could you please make sure that
- the tests are passing both with `MBEDTLS_RSA_NO_CRT` turned on and with `MBEDTLS_RSA_NO_CRT` turned off.
- the source code follows the [Mbed TLS coding standard](https://tls.mbed.org/kb/development/mbedtls-coding-standards) ?

**Comment by lubing521** [comment]
> @lubing521 while we are waiting for reviews, could you please make sure that
> 
> * the tests are passing both with `MBEDTLS_RSA_NO_CRT` turned on and with `MBEDTLS_RSA_NO_CRT` turned off.
It work well In my usage scenario when turn on mbedtls with CRT，This patch only works under the influence of the macro control MBEDTLS_RSA_NO_CRT.


> * the source code follows the [Mbed TLS coding standard](https://tls.mbed.org/kb/development/mbedtls-coding-standards) ?

I can't found any souce code not flow Mbed TLS coding standard now.

File: library/rsa.c [hunk 1]
```diff
@@ -207,11 +223,22 @@ static int rsa_check_context( mbedtls_rsa_context const *ctx, int is_priv,
     if( is_priv && mbedtls_mpi_cmp_int( &ctx->D, 0 ) <= 0 )
         return( MBEDTLS_ERR_RSA_BAD_INPUT_DATA );
 #else
-    if( is_priv &&
-        ( mbedtls_mpi_cmp_int( &ctx->DP, 0 ) <= 0 ||
-          mbedtls_mpi_cmp_int( &ctx->DQ, 0 ) <= 0  ) )
+    if(rsa_check_crt_parameters(ctx))
```
Comment (yanesca) [inline library/rsa.c:226]: > I can't found any souce code not flow Mbed TLS coding standard now.

I haven't reviewed the PR for style, but for example:
```suggestion
    if( rsa_check_crt_parameters( ctx ) )
```

Comment (lubing521) [reply library/rsa.c]: > > I can't found any souce code not flow Mbed TLS coding standard now.
> 
> I haven't reviewed the PR for style, but for example:

**Comment by lubing521** [comment]
@yanesca 
Hi 
how can I get this failure information as below show?
PR-4721-head TLS Testing — Failures: Windows-2013 Windows-mingw all_sh-freebsd-test_clang_opt all_sh-freebsd-test_cmake_out_of_source all_sh-freebsd-test_cmake_shared…

**Comment by yanesca** [comment]
> how can I get this failure information as below show?
PR-4721-head TLS Testing — Failures: Windows-2013 Windows-mingw all_sh-freebsd-test_clang_opt all_sh-freebsd-test_cmake_out_of_source all_sh-freebsd-test_cmake_shared…

Unfortunately, that CI is not public yet and one of the maintainers needs to post the relevant parts of the CI log here. (We should do that without asking, but feel free to remind us.)

In https://github.com/ARMmbed/mbedtls/pull/4721#issuecomment-869633968 I did just that: at 8734036 the CI was failing, because the RSA test suite was failing in builds where `MBEDTLS_RSA_NO_CRT` was turned off.

**Comment by yanesca** [comment]
We prefer rebasing to merging to resolve conflicts and use merge only as a last resort. Could you please remove the merge commit and rebase your commit on top of development instead?

**Comment by yanesca** [comment]
@lubing521 I think there was a misunderstanding, I was trying to ask that you remove the merge commit. You have added one and the PR has now two merge commits, could you please remove both merge commits and resolve the conflict by rebasing?

One way to do it is:
```
git reset HEAD~2
git reset --hard
git clean -df
git rebase upstream/development
```

At which point the conflicts can be resolved the same way as they were before. (Of course the history will have changed and will need to force push - `git push -f` to update the PR.)

**Comment by lubing521** [comment]
> @lubing521 I think there was a misunderstanding, I was trying to ask that you remove the merge commit. You have added one and the PR has now two merge commits, could you please remove both merge commits and resolve the conflict by rebasing?
> 
> One way to do it is:
> 
> ```
> git reset HEAD~2
> git reset --hard
> git clean -df
> git rebase upstream/development
> ```
> 
> At which point the conflicts can be resolved the same way as they were before. (Of course the history will have changed and will need to force push - `git push -f` to update the PR.)

sorry，i make mistake.
I push this patch on other way.

https://github.com/ARMmbed/mbedtls/pull/4742

**Comment by yanesca** [comment]
Thank you for deleting the merge commits!

Closing down in favour of #4742.

**Comment by yanesca** [comment]
@lubing521 Sorry, the other day I thought I understand it, but reading it back I am lost again.

>Keymaster TA in function TA_generate_key which on flie https://github.com/linaro-swg/kmgk/blob/master/keymaster/ta/generator.c
This function will create rsa key pair with CRT,so in optee-os with libmbedtls need enable CRT together.
if not ,this function will be called fail in rsa_check_context.
call stack as below in optee-os
syscall_asymm_operate--》crypto_acipher_rsanopad_decrypt–》mbedtls_rsa_private–》rsa_check_context

You write that you need to enable CRT because otherwise this part of the code fails:

```
#if !defined(MBEDTLS_RSA_NO_CRT)
    /* Modular exponentiation for P and Q is only
     * used for private key operations and if CRT
     * is used. */
    if( is_priv &&
        ( mbedtls_mpi_cmp_int( &ctx->P, 0 ) <= 0 ||
          mbedtls_mpi_get_bit( &ctx->P, 0 ) == 0 ||
          mbedtls_mpi_cmp_int( &ctx->Q, 0 ) <= 0 ||
          mbedtls_mpi_get_bit( &ctx->Q, 0 ) == 0  ) )
    {
        return( MBEDTLS_ERR_RSA_BAD_INPUT_DATA );
    }
#endif /* !MBEDTLS_RSA_NO_CRT */
```

If CRT is not enabled, then the option `MBEDTLS_RSA_NO_CRT` is defined and consequently the condition `!defined(MBEDTLS_RSA_NO_CRT)` will be false and the above code will not be present in the build and the program can't fail there.

What am I missing?

**Comment by lubing521** [comment]
@yanesca 
I think it needs to be reorganized

In our usage scenario,we need enable mbedtls in optee-os,but mbedtls no support CRT by default.
so we found that keymaster vts one case run fail  as show below 
```
130|console:/data/nativetest/VtsHalKeymasterV3_0TargetTest #
./VtsHalKeymasterV3_0TargetTest --gtest_filter=PerInstance/AttestationTest.RsaAttestation/0_default
Note: Google Test filter = PerInstance/AttestationTest.RsaAttestation/0_default
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from PerInstance/AttestationTest
[ RUN      ] PerInstance/AttestationTest.RsaAttestation/0_default
E/TC:? 0
E/TC:? 0 TA panicked with code 0xffff0001
E/LD:  Status of TA b13a05fa-cfd9-4c8e-9409-db2118d19a2f
E/LD:   arch: aarch64
E/LD:  region  0: va 0x40004000 pa 0x81180000 size 0x002000 flags rw-s (ldelf)
E/LD:  region  1: va 0x40006000 pa 0x81182000 size 0x009000 flags r-xs (ldelf)
E/LD:  region  2: va 0x4000f000 pa 0x8118b000 size 0x001000 flags rw-s (ldelf)
E/LD:  region  3: va 0x40010000 pa 0x8118c000 size 0x004000 flags rw-s (ldelf)
E/LD:  region  4: va 0x40014000 pa 0x81190000 size 0x001000 flags r--s
E/LD:  region  5: va 0x40015000 pa 0x81234000 size 0x009000 flags rw-s (stack)
E/LD:  region  6: va 0x4001e000 pa 0xf07fd4a8 size 0x003000 flags rw-- (param)
E/LD:  region  7: va 0x40021000 pa 0xf15e04a8 size 0x002000 flags rw-- (param)
E/LD:  region  8: va 0x40084000 pa 0x00001000 size 0x050000 flags r-xs [0]
E/LD:  region  9: va 0x400d4000 pa 0x00051000 size 0x053000 flags rw-s [0]
E/LD:   [0] b13a05fa-cfd9-4c8e-9409-db2118d19a2f @ 0x40084000
E/LD:  Call stack:
E/LD:   0x0000000040097800
E/LD:   0x000000004008cd8c
E/LD:   0x000000004008f714
E/LD:   0x0000000040085308
E/LD:   0x0000000040086b20
E/LD:   0x000000004009c0a0
E/LD:   0x0000000040094e08
hardware/interfaces/keymaster/3.0/vts/functional/keymaster_hidl_hal_test.cpp:3929: Failure
Expected: (cert_chain.size()) >= (2U), actual: 0 vs 2
Segmentation fault
[  261.709425] init: Untracked pid 5343 exited with status 0
[  261.717185] init: Untracked pid 5345 exited with status 0
139|console:/data/nativetest/VtsHalKeymasterV3_0TargetTest #
```
as a result  that Keymaster TA in function TA_generate_key which on flie https://github.com/linaro-swg/kmgk/blob/master/keymaster/ta/generator.c  This function will create rsa key pair with CRT,
but mbedtls not support CRT by default , this will casues TA panic.
 
After analysis we need the following code support
```
#if !defined(MBEDTLS_RSA_NO_CRT)
    /*
     * DP = D mod (P - 1)
     * DQ = D mod (Q - 1)
     * QP = Q^-1 mod P
     */
    MBEDTLS_MPI_CHK( mbedtls_rsa_deduce_crt( &ctx->P, &ctx->Q, &ctx->D,
                                             &ctx->DP, &ctx->DQ, &ctx->QP ) );
#endif /* MBEDTLS_RSA_NO_CRT */
```
so we enable CRT in the file lib/libmbedtls/include/mbedtls_config_kernel.h by  Comment line
#define MBEDTLS_RSA_NO_CRT 

keymaster vts work well with optee-os after enable mbedtls with CRT.

But something else happened, But the xtest program, a companion to optee-os, doesn't work.

xtest 4006 for rsa test fail .

```
# xtest 4006
Test ID: 4006
Run test suite with level=0

TEE test application started over default TEE instance
######################################################
#
# regression
#
######################################################
 
* regression_4006 Test TEE Internal API Asymmetric Cipher operations
o regression_4006.1 Asym Crypto case 0 algo 0x60000030 line 2728
regression_4000.c:3900: ta_crypt_cmd_asymmetric_decrypt(c, &session, op, ((void *)0), 0, out_enc, out_enc_size, out, &out_size) has an unexpected value: 0xffff0006 = TEEC_ERROR_BAD_PARAMETERS, expected 0x0 = TEEC_SUCCESS
  regression_4006.1 FAILED
  regression_4006 FAILED
+-----------------------------------------------------
Result of testsuite regression filtered by "4006":
regression_4006.1 FAILED first error at regression_4000.c:3900
regression_4006 FAILED
+-----------------------------------------------------
36 subtests of which 1 failed
1 test case of which 1 failed
91 test cases were skipped
TEE test application done!


D/TC:? 0 tee_ta_init_pseudo_ta_session:299 Lookup pseudo TA cb3e5ba0-adf1-11e0-998b-0002a5d5c51b
D/TC:? 0 ldelf_load_ldelf:91 ldelf load address 0x40006000
D/LD:  ldelf:134 Loading TS cb3e5ba0-adf1-11e0-998b-0002a5d5c51b
D/TC:? 0 ldelf_syscall_open_bin:142 Lookup user TA ELF cb3e5ba0-adf1-11e0-998b-0002a5d5c51b (Secure Storage TA)
D/TC:? 0 ldelf_syscall_open_bin:146 res=0xffff0008
D/TC:? 0 ldelf_syscall_open_bin:142 Lookup user TA ELF cb3e5ba0-adf1-11e0-998b-0002a5d5c51b (REE)
D/TC:? 0 ldelf_syscall_open_bin:146 res=0
D/LD:  ldelf:168 ELF (cb3e5ba0-adf1-11e0-998b-0002a5d5c51b) at 0x40079000
D/TC:? 0 tee_ta_invoke_command:823 Error: ffff0006 of 4
D/TC:? 0 tee_ta_close_session:512 csess 0x9a115530 id 1
D/TC:? 0 tee_ta_close_session:531 Destroy session
D/TC:? 0 destroy_context:308 Destroy TA ctx (0x9a1154a0)
 ```

I found that xtest 4006 will test without crt parameters and with crt parameters.

```
static const struct xtest_ac_case xtest_ac_cases[] = {

/* RSA test without crt parameters */

XTEST_AC_RSA_CASE(0, TEE_ALG_RSA_NOPAD, TEE_MODE_ENCRYPT, XTEST_AC_RSA_CASE(0, TEE_ALG_RSA_NOPAD, TEE_MODE_ENCRYPT,   ac_rsassa_vect1, NULL_ARRAY, WITHOUT_SALT),

....

...

/* RSA test with crt parameters */

XTEST_AC_RSA_CASE(0, TEE_ALG_RSA_NOPAD, TEE_MODE_ENCRYPT,XTEST_AC_RSA_CASE(0, TEE_ALG_RSA_NOPAD, TEE_MODE_ENCRYPT,   ac_rsassa_vect1, ARRAY, WITHOUT_SALT),

....

...

}
```
By comparing the tomcrypt library ,we get some useful information as below

tomcrypt and mbedtls in optee-os deal with diffrence ways.

1.tomcrypt will do mp_exptmod when xtest 4006 without crt parameters,so no affect after test process.

2.mbedtls withh CRT will check CRT parameters in function rsa_check_context, it will return MBEDTLS_ERR_RSA_BAD_INPUT_DATA when xtest 4006 without crt parameters, so xtest 4006 run fail.

call stack as below :
syscall_asymm_operate--》crypto_acipher_rsanopad_decrypt–》mbedtls_rsa_private–》rsa_check_context

 ```
#if !defined(MBEDTLS_RSA_NO_CRT)
    /* Modular exponentiation for P and Q is only
     * used for private key operations and if CRT
     * is used. */
    if( is_priv &&
        ( mbedtls_mpi_cmp_int( &ctx->P, 0 ) <= 0 ||
          mbedtls_mpi_get_bit( &ctx->P, 0 ) == 0 ||
          mbedtls_mpi_cmp_int( &ctx->Q, 0 ) <= 0 ||
          mbedtls_mpi_get_bit( &ctx->Q, 0 ) == 0  ) )
    {
        return( MBEDTLS_ERR_RSA_BAD_INPUT_DATA );
    }
#endif /* !MBEDTLS_RSA_NO_CRT */
```

To solve this scenario we borrowed from the tomcrypt implementation,

RSA support CRT and NO_CRT mode at same time when mbedtls enable CRT.

**Pull request closed without merging**